File: C++/Subsets_II.cpp
class Solution {
public:
    /*
    Recursion Function:
    For a existing subset s, there are k distinct unselected numbers.
    Generate k new different subsets by copying s and append a candidate number.
    For every new subset, call recursion function if this subset can generate new subsets.
    */
    vector<vector<int> > ret;
    vector<vector<int> > subsetsWithDup(vector<int> &S) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        ret.clear();
        vector<int> cand;
        ret.push_back(cand);
        if (S.size() == 0) return ret;
        sort(S.begin(), S.end());
        solve(S, cand, 0);
        return ret;
    }
    void solve(vector<int> &S, vector<int> &cand, int minIndex){
        if (minIndex >= S.size()) return;
        for (int n = minIndex; n < S.size(); n ++){
            if ((n > minIndex && S[n] != S[n-1]) || (n == minIndex)){
                vector<int>newCand(cand);
                newCand.push_back(S[n]);
                ret.push_back(newCand);
                if (n+1 < S.size()){
                    solve(S, newCand, n+1);
                }
            }
        }
    }
};

File: Java/Subsets_II.java
public class Solution {
    /*
    Recursion Function:
    For a existing subset s, there are k distinct unselected numbers.
    Generate k new different subsets by copying s and append a candidate number.
    For every new subset, call recursion function if this subset can generate new subsets.
    */
    ArrayList<ArrayList<Integer>> ret = new ArrayList<ArrayList<Integer>>();
    public ArrayList<ArrayList<Integer>> subsetsWithDup(int[] num) {
        ret.clear();
        Arrays.sort(num);
        ArrayList<Integer> cand = new ArrayList<Integer>();
        ret.add(cand);
        solve(cand, num, 0);
        return ret;
    }
    
    private void solve (ArrayList<Integer> cand, int[] num, int start) {
        for (int i = start; i < num.length; i ++) {
            if (i == start || (i > start && num[i] != num[i-1])){
                ArrayList<Integer> newCand = new ArrayList<Integer> (cand);
                newCand.add(num[i]);
                ret.add(newCand);
                if (i < num.length - 1) {
                    solve (newCand, num, i+1);
                }
            }
        }
    }
}
